src/azure-cli-core/azure/cli/core/tests/test_keys.py

```diff
@@ -0,0 +1,135 @@
+
+    def test_public_key_file_exists_no_permissions(self):
+        # Create public key file with no read or write access
+        public_key_path = self._create_new_temp_key_file(self.public_key)
+        os.chmod(public_key_path, 0o000)
+
+        # Check that CLIError exception is raised when generate_ssh_keys is called.
+        with self.assertRaises(CLIError):
+            generate_ssh_keys("", public_key_path)
+
+    def test_private_key_file_exists_no_permissions(self):
+        # Create private key file with no read or write access
+        private_key_path = self._create_new_temp_key_file(self.private_key)
+        os.chmod(private_key_path, 0o000)
+
+        # Check that CLIError exception is raised when generate_ssh_keys is called.
+        with self.assertRaises(CLIError):
+            public_key_path = private_key_path + ".pub"
+            generate_ssh_keys(private_key_path, public_key_path)
+
+    def test_private_key_file_exists_encrypted(self):
+        # Create empty private key file
+        private_key_path = self._create_new_temp_key_file("")
+
+        # Write encrypted key into file
+        self.key.write_private_key_file(private_key_path, "test")
+
+        # Check that CLIError exception is raised when generate_ssh_keys is called.
+        with self.assertRaises(CLIError):
+            public_key_path = private_key_path + ".pub"
+            generate_ssh_keys(private_key_path, public_key_path)
+
+    def test_private_key_file_exists(self):
+        # Create private key file
+        private_key_path = self._create_new_temp_key_file(self.private_key)
+
+        # Call generate_ssh_keys and assert that returned public key same as original
+        public_key_path = private_key_path + ".pub"
+        new_public_key = generate_ssh_keys(private_key_path, public_key_path)
+        self.assertEqual(self.public_key, new_public_key)
+
+        # Check that correct public key file has been created
+        with open(public_key_path, 'r') as f:
+            public_key = f.read()
+            self.assertEqual(self.public_key, public_key)
+
+        # Check that private key file contents unchanged
+        with open(private_key_path, 'r') as f:
+            private_key = f.read()
+            self.assertEqual(self.private_key, private_key)
+
+    def test_private_key_file_new(self):
+        # create random temp file name
+        f = tempfile.NamedTemporaryFile(mode='w', dir=self._tempdirName)
+        f.close()
+        private_key_path = f.name
+
+        # Call generate_ssh_keys and assert that returned public key same as original
+        public_key_path = private_key_path + ".pub"
+        new_public_key = generate_ssh_keys(private_key_path, public_key_path)
+
+        # Check that public key returned is same as public key in public key path
+        with open(public_key_path, 'r') as f:
+            public_key = f.read()
+            self.assertEqual(public_key, new_public_key)
```

Comment:
Just a question: so you are sure that the keys generated at different time in the same machine would be the same?

Reply:
So, this method asserts that the `new_public_key` returned by `generate_ssh_keys` is the same as the one that the method wrote to the public key file it created under `public_key_path`.

Reply:
Thanks for all the feedback, btw.